astroid/context.py

```diff
@@ -81,6 +85,31 @@ def __init__(self, path=None, inferred=None):
         for call arguments
         """
 
+    @property
+    def nodes_inferred(self):
+        """
+        Number of nodes inferred in this context and all its clones/decendents
+
+        Wrap inner value in a mutable cell to allow for mutating a class
+        variable in the presence of __slots__
+        """
+        return self._nodes_inferred[0]
+
+    @nodes_inferred.setter
+    def nodes_inferred(self, value):
+        self._nodes_inferred[0] = value
+
+    @property
+    def inferred(self):
+        """
+        :type: optional[NodeNG]
+
+        The bound node of the given context
+
+        e.g. the bound node of object.__new__(cls) is the object node
+        """
+        return _INFERENCE_CACHE
```

Comment:
If i understand well, here we will store every inferred node in a global variable that will exists throughout the life of the program. 
It will definitely allow cpu time sparing but what about memory consumption?

Reply:
Is the number of inferred node limit sufficient for the speed up ? There are really big code bases using pylint and having a cache can affects them to the point it's not possible to run pylint on them at all.

Reply:
> If i understand well, here we will store every inferred node in a global variable that will exists throughout the life of the program.
> It will definitely allow cpu time sparing but what about memory consumption?

Your understanding is correct. It could theoretically grow without bound, but I suppose many other parts of astroid (such as `astroid.MANAGER.astroid_cache`) can as well. I regularly see pylint/astroid runs in the hundreds of MiB, but without a benchmark, I'm not sure there's a meaningful way to answer this question.

I tried running pylint (commit 9361fa0630ce920081b57865be7882a39e9da1e3) on itself with and without these astroid changes.

With `wip/performance` (0d7d7989458326a863f7fa942e662c5446d170e1):

```
26.77user 0.25system 0:27.08elapsed 99%CPU (0avgtext+0avgdata 231516maxresident)k
3312inputs+248outputs (3major+81935minor)pagefaults 0swaps
```

Without (d3fea8fee322200884920620cd9f9cb339ee9158):

```
45.15user 0.24system 0:45.43elapsed 99%CPU (0avgtext+0avgdata 235892maxresident)k
0inputs+336outputs (0major+83907minor)pagefaults 0swaps
```

So memory usage (max RSS) went down from 235MB -> 231MB


> Is the number of inferred node limit sufficient for the speed up ? There are really big code bases using pylint and having a cache can affects them to the point it's not possible to run pylint on them at all.

They address different problems: the inferred node limit is to stop pathological cases, like really large files. The inferred node cache helps with re-use, like when scanning a package with a lot of files. The main problem I saw with the inference cache was that it was re-built from scratch for every node so the hit-rate was very low, and cloning it introduced a large number of small copies which was inefficient time-wise. In fact, it was so inefficient that just removing it altogether produces a small but noticeable (~2%) speed-up on astroid 2.5.6 (middle of 3 runs):

astroid-2.5.6 (36dda3fc8a5826b19a33a0ff29402b61d6a64fc2):
```
38.57user 0.26system 0:38.86elapsed 99%CPU (0avgtext+0avgdata 231020maxresident)k
0inputs+72outputs (0major+81740minor)pagefaults 0swaps
```

With no-cache.patch applied:
```
37.77user 0.25system 0:38.05elapsed 99%CPU (0avgtext+0avgdata 230788maxresident)k
0inputs+72outputs (0major+81551minor)pagefaults 0swaps
```

`no-cache.patch`:
```patch
diff --git a/astroid/context.py b/astroid/context.py
index 18220ec..73f1f4d 100644
--- a/astroid/context.py
+++ b/astroid/context.py
@@ -27,11 +27,10 @@ class InferenceContext:
         "lookupname",
         "callcontext",
         "boundnode",
-        "inferred",
         "extra_context",
     )
 
-    def __init__(self, path=None, inferred=None):
+    def __init__(self, path=None):
         self.path = path or set()
         """
         :type: set(tuple(NodeNG, optional(str)))
@@ -64,14 +63,6 @@ class InferenceContext:
 
         e.g. the bound node of object.__new__(cls) is the object node
         """
-        self.inferred = inferred or {}
-        """
-        :type: dict(seq, seq)
-
-        Inferred node contexts to their mapped results
-        Currently the key is ``(node, lookupname, callcontext, boundnode)``
-        and the value is tuple of the inferred results
-        """
         self.extra_context = {}
         """
         :type: dict(NodeNG, Context)
@@ -102,7 +93,7 @@ class InferenceContext:
         starts with the same context but diverge as each side is inferred
         so the InferenceContext will need be cloned"""
         # XXX copy lookupname/callcontext ?
-        clone = InferenceContext(self.path, inferred=self.inferred)
+        clone = InferenceContext(self.path)
         clone.callcontext = self.callcontext
         clone.boundnode = self.boundnode
         clone.extra_context = self.extra_context
diff --git a/astroid/node_classes.py b/astroid/node_classes.py
index 7faf681..089b579 100644
--- a/astroid/node_classes.py
+++ b/astroid/node_classes.py
@@ -358,11 +358,6 @@ class NodeNG:
             yield from self._infer(context, **kwargs)
             return
 
-        key = (self, context.lookupname, context.callcontext, context.boundnode)
-        if key in context.inferred:
-            yield from context.inferred[key]
-            return
-
         generator = self._infer(context, **kwargs)
         results = []
 
@@ -378,7 +373,6 @@ class NodeNG:
 
         # Cache generated results for subsequent inferences of the
         # same node using the same context
-        context.inferred[key] = tuple(results)
         return
 
     def _repr_name(self):
```